src/core/net/srp_server.cpp

```diff
@@ -381,26 +382,26 @@ bool Server::HasNameConflictsWith(Host &aHost) const
 
 void Server::HandleServiceUpdateResult(ServiceUpdateId aId, Error aError)
 {
-    UpdateMetadata *update = mOutstandingUpdates.FindMatching(aId);
+    UpdateMetadata *update = mOutstandingUpdates.RemoveMatching(aId);
 
-    if (update != nullptr)
-    {
-        HandleServiceUpdateResult(update, aError);
-    }
-    else
+    if (update == nullptr)
     {
         LogInfo("Delayed SRP host update result, the SRP update has been committed (updateId = %lu)", ToUlong(aId));
+        ExitNow();
     }
-}
 
-void Server::HandleServiceUpdateResult(UpdateMetadata *aUpdate, Error aError)
-{
-    LogInfo("Handler result of SRP update (id = %lu) is received: %s", ToUlong(aUpdate->GetId()),
-            ErrorToString(aError));
+    update->SetError(aError);
+
+    LogInfo("Handler result of SRP update (id = %lu) is received: %s", ToUlong(update->GetId()), ErrorToString(aError));
 
-    IgnoreError(mOutstandingUpdates.Remove(*aUpdate));
-    CommitSrpUpdate(aError, *aUpdate);
-    aUpdate->Free();
+    // We add new `update` at the tail of the `mCompletedUpdates` list
+    // so that updates are processed in the order we receive the
+    // `HandleServiceUpdateResult()` callbacks for them. The
+    // completed updates are processed from `mCompletedUpdateTask`
+    // and `ProcessCompletedUpdates()`.
+
+    mCompletedUpdates.PushAfterTail(*update);
+    mCompletedUpdateTask.Post();
```

Comment:
We may don't need a Tasklet.

The signature of `otSrpServerServiceUpdateHandler` should clarify that `aHost` is guaranteed to be available only before the handler callback returns (so the upper layer shouldn't save a reference to `aHost` and use it after the callback returns. Instead, use the `aId`).

```
typedef void (*otSrpServerServiceUpdateHandler)(otSrpServerServiceUpdateId aId,
                                                const otSrpServerHost     *aHost,
                                                uint32_t                   aTimeout,
                                                void                      *aContext);
```

That means we can cleanup / process the `mCompletedUpdates` right after the service update handler has been invoked. Like this:

```
void Server::InformUpdateHandlerOrCommit(Error aError, Host &aHost, const MessageMetadata &aMetadata)
{
// ...

            LogInfo("SRP update handler is notified (updatedId = %lu)", ToUlong(update->GetId()));
            mServiceUpdateHandler.Invoke(update->GetId(), &aHost, static_cast<uint32_t>(kDefaultEventsHandlerTimeout));
            **ProcessCompletedUpdates();**
            ExitNow();

// ...
}
```

This makes the lifecycle of `aHost` clearer given we don't know exactly when the Tasklet will be be process exatcly.

Reply:
@wgtdkp reading the comment I think I may have misunderstood it.  😄 

> The signature of otSrpServerServiceUpdateHandler should clarify that aHost is guaranteed to be available only before the handler callback returns (so the upper layer shouldn't save a reference to aHost and use it after the callback returns. Instead, use the aId)

I think by "handler callback" you mean `otSrpServerServiceUpdateHandler` itself? 
I use "callback" to refer to `otSrpServerHandleServiceUpdateResult()` call reporting status back to SRP server.

So If I not mistaken, we are saying the same thing, i.e. the `aHost` should remain valid until `otSrpServerServiceUpdateHandler` call returns (independent of `otSrpServerHandleServiceUpdateResult()` is called).

----------

> That means we can cleanup / process the mCompletedUpdates right after the service update handler has been invoked.

Regarding suggestion of not using `Tasklet` and instead calling `ProcessCompletedUpdates()` after the `mServiceUpdateHandler.Invoke()`, it will work for the case where the `ServiceUpdateResult()` callback is called  from `mServiceUpdateHandler` call itself, but what if it is called afterwards?  
- To make this work (without `Tasklet`) we need a way to track if we are in middle of `mServiceUpdateHandler` call and then behave differently in `ServiceUpdateResult()`. 
- `Tasklet` seems to be an easier way to de-couple these (and it is the pattern used in other places in OT code).

Reply:
In general, I think callbacks should be called after the request API has completed. Tasklets is a convenient way to implement such behavior.

Reply:
> I think by "handler callback" you mean otSrpServerServiceUpdateHandler itself?

Yes.

> Regarding suggestion of not using Tasklet and instead calling ProcessCompletedUpdates() after the mServiceUpdateHandler.Invoke(), it will work for the case where the ServiceUpdateResult() callback is called from mServiceUpdateHandler call itself, but what if it is called afterwards?

yeah, I missed this. Agree it's easier to use Tasklet to handle both cases. Thanks @abtink!